Leverage First-class callable syntax

tigitz opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | no 
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | -
| License       | MIT
| Doc PR        | -

### Rationale
https://wiki.php.net/rfc/first_class_callable_syntax

Mainly: 
>  The advantage is that the new syntax is accessible to static analysis, and respects the scope at the point where the callable is created.

I'd argue that it also improves readability and IDE color syntax also helps:

![image](https://user-images.githubusercontent.com/1524501/191912084-7ee933c5-dda1-4176-86f1-cd6511c58aa4.png)

I've manually reviewed each changes and discarded some of them where `[Foo::class, 'method']` was intended to be tested with this specific syntax

tigitz commented:
There are related failures in the test suite, I'm looking into it.

stof commented on src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php:
```diff
@@ -194,7 +194,7 @@ public function shutdown()
         restore_error_handler();
 
-        if ($currErrorHandler !== [$this, 'handleError']) {
+        if ($currErrorHandler !== $this->handleError(...)) {
```
Are you sure this works ? To me, this will create a *new* Closure object and so it won't be `===`

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -764,7 +764,7 @@ private function registerHttpCacheConfiguration(array $config, ContainerBuilder
             $container->getDefinition('http_cache')
                   ->addArgument((new Definition('void'))
-                      ->setFactory([Request::class, 'enableHttpMethodParameterOverride'])
+                      ->setFactory(Request::enableHttpMethodParameterOverride(...))
```
@nicolas-grekas Does the DI component supports dumping the container when using first-class callables for a factory ? I don't remember whether this has been added (and if yes, in which version), as the dumper does not sure arbitrary Closure.

but otherwise, I would be fine reverting that for all those `setFactory` calls. `setFactory` does not expect a callable. It expects a definition for producing a callable.

tigitz replied:
AFAIU it doesn't support Closures so I'm reverting these changes. I'll open an issue to keep track of this potential improvement.

stof commented on src/Symfony/Component/ErrorHandler/ErrorHandler.php:
```diff
@@ -135,18 +135,18 @@ public static function register(self $handler = null, bool $replace = true): sel
             $handlerIsRegistered = true;
         }
-        if (\is_array($prev = set_exception_handler([$handler, 'handleException'])) && $prev[0] instanceof self) {
+        if (\is_array($prev = set_exception_handler($handler->handleException(...))) && $prev[0] instanceof self) {
```
I'm quite sure those `$prev[0] instanceof self` checks won't work anymore if you use a first-class callable instead of the array-based callables.

tigitz replied:
You're right, I'm reverting all changes related to the ErrorHandler component anyway :+1:

chalasr commented on src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php:
```diff
@@ -194,7 +194,7 @@ public function shutdown()
         restore_error_handler();
 
-        if ($currErrorHandler !== [$this, 'handleError']) {
+        if ($currErrorHandler !== $this->handleError(...)) {
```
The PHPUnit bridge cannot use this feature as it supports PHP 7.1 for testing purposes (the lowest supported php version of the lowest maintained Symfony branch):
https://github.com/symfony/symfony/blob/7b212a11be00e080b2416e6f563cd5349df1a54f/src/Symfony/Bridge/PhpUnit/composer.json#L19

tigitz replied:
Ha, good catch thanks. Still learning all the little quirks here and there. Lovin' it :)

tigitz commented:
So I've dug a little more to properly implement these improvements. Hopefully, I got a good enough understanding now.

Mainly I've been reminded that
```php
$this->foo(...) === $this->foo(...) // false
[$this, 'foo'] === [$this, 'foo'] // true
```
Even though both are callables, they are `Closure` and `array` respectively. So switching one with the other is not as straightforward as I previously thought.

This has consequences on unit tests where equality is tested. Symfony public api returning callables with the array syntax form must be kept that way and therefore their related tests too.

**A couple more points:**

- As stated by @stof `\Symfony\Component\DependencyInjection\Definition::setFactory` doesn't support closures.

- I refrained from applying this in `\Symfony\Component\EventDispatcher\EventDispatcher::addSubscriber` 
```diff
--- a/src/Symfony/Component/EventDispatcher/EventDispatcher.php
+++ b/src/Symfony/Component/EventDispatcher/EventDispatcher.php
@@ -163,10 +163,10 @@ class EventDispatcher implements EventDispatcherInterface
             if (\is_string($params)) {
                 $this->addListener($eventName, [$subscriber, $params]);
             } elseif (\is_string($params[0])) {
-                $this->addListener($eventName, [$subscriber, $params[0]], $params[1] ?? 0);
+                $this->addListener($eventName, $subscriber->{$params[0]}(...), $params[1] ?? 0);
             } else {
                 foreach ($params as $listener) {
-                    $this->addListener($eventName, [$subscriber, $listener[0]], $listener[1] ?? 0);
+                    $this->addListener($eventName, $subscriber->{$params[0]}(...), $listener[1] ?? 0);
                 }
             }
         }
```

As it changes the expected behavior of what's stored inside the `listeners` (closure instead of array) when adding a `Subscriber`. 
Even though it's valid, as the param is typed `array|callable`, it could break userland code if they implicitly rely on having an array here when `addSubscriber` is called.

- I've reverted updates on `EventDispatcherTest::testGetLazyListeners` , `testRemoveFindsLazyListeners`, `testPriorityFindsLazyListeners` as it fails and I don't understand what it's actually testing to fix them. Maybe @nicolas-grekas can provide some guidance here :pray: 
- I've reverted updates on the `ErrorHandler` component as it breaks a lot of tests and there's too much dark magic going on to make the modernizing effort worth it.
- It seems `\Symfony\Component\HttpKernel\Controller\ArgumentResolver\ServiceValueResolver::resolve` doesn't support `$request->attributes->get('_controller')` returning a `Closure`. TBC.

Ready to be reviewed again. Test failures are not related.

nicolas-grekas commented on src/Symfony/Component/EventDispatcher/Tests/EventDispatcherTest.php:
```diff
@@ -398,7 +398,7 @@ public function testMutatingWhilePropagationIsStopped()
         $this->assertFalse($test->preFooInvoked);
 
-        $this->assertsame(0, $this->dispatcher->getListenerPriority('foo', [$test, 'preFoo']));
+        $this->assertsame(0, $this->dispatcher->getListenerPriority('foo', $listenerCall));
```
Using the syntax here should work instead of passing the exact same closure instance.

More importantly, we should keep testing that the previous syntax works in all test cases of this file.
Instead of changing existing tests, we would add more that leverage the new syntax (to test that the ED can effectively compare two different instances of closures for eg, as should be the case here.)

We added some of these in #46262, maybe that's enough.

tigitz replied:
Reverted so that all tests are using the previous array callable syntax at least once.

I tried to mix the rest with first class callable syntax to widen the scope of what's tested when it made sense. Consequently I had to use `assertEquals` in some places to adapt to closure equality comparison.

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Tests/DataCollector/RequestDataCollectorTest.php:
```diff
@@ -102,7 +102,7 @@ public function provideControllerCallables()
             [
                 '"Regular" callable',
-                [$this, 'testControllerInspection'],
+                $this->testControllerInspection(...),
```
looks like this was on purpose - the next item tests with a closure

tigitz replied:
Reverted and added a test case for `First-class callable closure`

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Tests/DataCollector/RequestDataCollectorTest.php:
```diff
@@ -135,7 +135,7 @@ function () { return 'foo'; },
             [
                 'Static callable with instance',
-                [$this, 'staticControllerMethod'],
+                $this->staticControllerMethod(...),
```
was on purpose to me

tigitz replied:
Reverted

nicolas-grekas reviewed (commented):
> Mainly I've been reminded that
> ```php
> $this->foo(...) === $this->foo(...) // false
> [$this, 'foo'] === [$this, 'foo'] // true
> ```

Note this also: `$this->foo(...) == $this->foo(...) // true`, see #46262

> As stated by @stof `\Symfony\Component\DependencyInjection\Definition::setFactory` doesn't support closures.

It might but we first need to teach PhpDumper how to dump such closures. There might be an issue with closures created out of non-public methods though.

> I refrained from applying this in `\Symfony\Component\EventDispatcher\EventDispatcher::addSubscriber` 

You did right, this would badly break BC.

> I've reverted updates on `EventDispatcherTest::testGetLazyListeners`

We test lazy-listeners there: those are not native php callables so they cannot use first-class callable syntax (they are `[\Closure, string]` tuples.)

> I've reverted updates on the `ErrorHandler` component as it breaks a lot of tests and there's too much dark magic going on to make the modernizing effort worth it.

No dark magic here, just regular php error handling ;) Yes, adopting a new syntax should not lead to invasive changes.

> It seems `ServiceValueResolver::resolve` doesn't support `$request->attributes->get('_controller')` returning a `Closure`.

It does: it skips them. That's expected: controllers-as-services are never dumped as closures anyway.

tigitz committed:
Leverage First-class callable syntax

tigitz commented:
@nicolas-grekas Thanks for your thorough review, very motivating :) 

Ready to be reviewed again. :+1:

nicolas-grekas commented:
Thank you @tigitz.
